firedrake/extrusion_utils.py

```diff
@@ -43,119 +46,156 @@ def make_extruded_coords(extruded_topology, base_coords, ext_coords,
-        if base_coords.ufl_domain().ufl_cell().topological_dimension() not in [1, 2]:
+        tdim = base_coords.ufl_domain().ufl_cell().topological_dimension()
+        if tdim not in [1, 2]:
             raise NotImplementedError("Hedgehog extrusion not implemented for %s" % base_coords.ufl_domain().ufl_cell())
-        kernel = op2.Kernel("""
-static inline void pyop2_kernel_radial_hedgehog_extrusion(double *ext_coords,
-                                                          const double *base_coords,
-                                                          const double *layer_height,
-                                                          int layer) {
-    double v0[%(base_coord_dim)d];
-    double v1[%(base_coord_dim)d];
-    double n[%(base_coord_dim)d];
-    double x[%(base_coord_dim)d] = {0};
-    double dot = 0.0;
-    double norm = 0.0;
-    int i, c, d;
-    if (%(base_coord_dim)d == 2) {
-        /*
-         * normal is:
-         * (0 -1) (x2 - x1)
-         * (1  0) (y2 - y1)
-         */
-        n[0] = -(base_coords[%(base_coord_dim)d+1] - base_coords[1]);
-        n[1] = base_coords[%(base_coord_dim)d] - base_coords[0];
-    } else if (%(base_coord_dim)d == 3) {
-        /*
-         * normal is
-         * v0 x v1
-         *
-         *    /\\
-         * v0/  \\
-         *  /    \\
-         * /------\\
-         *    v1
-         */
-        for (i = 0; i < 3; ++i) {
-            v0[i] = base_coords[%(base_coord_dim)d+i] - base_coords[i];
-            v1[i] = base_coords[2*%(base_coord_dim)d+i] - base_coords[i];
-        }
-        n[0] = v0[1] * v1[2] - v0[2] * v1[1];
-        n[1] = v0[2] * v1[0] - v0[0] * v1[2];
-        n[2] = v0[0] * v1[1] - v0[1] * v1[0];
-    }
-    for (i = 0; i < %(base_map_arity)d; ++i) {
-        for (c = 0; c < %(base_coord_dim)d; ++c) {
-            x[c] += base_coords[i*%(base_coord_dim)d+c];
-        }
-    }
-    for (i = 0; i < %(base_coord_dim)d; ++i) {
-        dot += x[i] * n[i];
-        norm += n[i] * n[i];
-    }
-    /*
-     * Make inward-pointing normals point out
-     */
-    norm = sqrt(norm);
-    norm *= (dot < 0 ? -1 : 1);
-    for (d = 0; d < %(base_map_arity)d; ++d) {
-        for (c = 0; c < %(base_coord_dim)d; ++c ) {
-            ext_coords[2*d*%(base_coord_dim)d+c] = base_coords[d*%(base_coord_dim)d+c] + n[c] * layer_height[0] * layer / norm;
-            ext_coords[(2*d+1)*%(base_coord_dim)d+c] = base_coords[d*%(base_coord_dim)d+c] + n[c] * layer_height[0] * (layer + 1)/ norm;
-        }
-    }
-}""" % {'base_map_arity': base_coords.cell_node_map().arity, 'base_coord_dim': base_coords.function_space().value_size},
-            "pyop2_kernel_radial_hedgehog_extrusion")
+        # tdim == 1:
+        #
+        # normal is:
+        # (0 -1) (x2 - x1)
+        # (1  0) (y2 - y1)
+        #
+        # tdim == 2:
+        # normal is
+        # v0 x v1
+        #
+        #    /\\
+        # v0/  \\
+        #  /    \\
+        # /------\\
```

Comment:
No need for double backslash in comment.

Reply:
Fixed